File: backend/app/schemas/study.py
from datetime import date, datetime

from pydantic import BaseModel, Field


class StudyCategoryCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    target_minutes: int = 0
    memo: str | None = None


class StudyCategoryUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=120)
    target_minutes: int | None = None
    memo: str | None = None


class StudyCategoryResponse(BaseModel):
    id: int
    name: str
    target_minutes: int
    memo: str | None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class StudySubjectCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    category_id: int | None = None
    exam_name: str | None = None
    target_minutes: int = 0
    color: str | None = None
    memo: str | None = None


class StudySubjectUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=120)
    category_id: int | None = None
    exam_name: str | None = None
    target_minutes: int | None = None
    color: str | None = None
    memo: str | None = None


class StudySubjectResponse(BaseModel):
    id: int
    name: str
    category_id: int | None
    category_name: str | None = None
    exam_name: str | None
    target_minutes: int
    color: str | None
    memo: str | None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class StudyLogCreate(BaseModel):
    subject_id: int
    studied_on: date
    started_at: datetime | None = None
    ended_at: datetime | None = None
    duration_minutes: int
    material: str | None = None
    unit: str | None = None
    method: str | None = None
    understanding: int | None = None
    memo: str | None = None


class StudyLogUpdate(BaseModel):
    subject_id: int | None = None
    studied_on: date | None = None
    started_at: datetime | None = None
    ended_at: datetime | None = None
    duration_minutes: int | None = None
    material: str | None = None
    unit: str | None = None
    method: str | None = None
    understanding: int | None = None
    memo: str | None = None


class StudyLogResponse(BaseModel):
    id: int
    subject_id: int
    subject_name: str
    category_id: int | None = None
    category_name: str | None = None
    studied_on: date
    started_at: datetime | None
    ended_at: datetime | None
    duration_minutes: int
    material: str | None
    unit: str | None
    method: str | None
    understanding: int | None
    memo: str | None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class StudySubjectSummary(BaseModel):
    subject_id: int
    subject_name: str
    category_id: int | None = None
    category_name: str | None = None
    exam_name: str | None
    target_minutes: int
    total_minutes: int
    log_count: int


class StudyCategorySummary(BaseModel):
    category_id: int | None
    category_name: str
    target_minutes: int
    subject_target_minutes: int
    effective_target_minutes: int
    total_minutes: int
    log_count: int
    subject_count: int


class StudyDailySummary(BaseModel):
    studied_on: date
    total_minutes: int


class StudySummaryResponse(BaseModel):
    total_minutes: int
    total_logs: int
    subject_target_minutes: int
    category_target_minutes: int
    effective_target_minutes: int
    category_summaries: list[StudyCategorySummary]
    subject_summaries: list[StudySubjectSummary]
    daily_summaries: list[StudyDailySummary]

File: frontend/src/api/study.ts
import { apiClient } from "./client";

export type StudySubject = {
  id: number;
  name: string;
  exam_name?: string | null;
  target_minutes: number;
  color?: string | null;
  memo?: string | null;
  created_at: string;
  updated_at: string;
};

export type StudySubjectInput = {
  name: string;
  exam_name?: string | null;
  target_minutes?: number;
  color?: string | null;
  memo?: string | null;
};

export type StudyLog = {
  id: number;
  subject_id: number;
  subject_name: string;
  studied_on: string;
  started_at?: string | null;
  ended_at?: string | null;
  duration_minutes: number;
  material?: string | null;
  unit?: string | null;
  method?: string | null;
  understanding?: number | null;
  memo?: string | null;
  created_at: string;
  updated_at: string;
};

export type StudyLogInput = {
  subject_id: number;
  studied_on: string;
  started_at?: string | null;
  ended_at?: string | null;
  duration_minutes: number;
  material?: string | null;
  unit?: string | null;
  method?: string | null;
  understanding?: number | null;
  memo?: string | null;
};

export type StudySubjectSummary = {
  subject_id: number;
  subject_name: string;
  exam_name?: string | null;
  target_minutes: number;
  total_minutes: number;
  log_count: number;
};

export type StudyDailySummary = {
  studied_on: string;
  total_minutes: number;
};

export type StudySummary = {
  total_minutes: number;
  total_logs: number;
  subject_summaries: StudySubjectSummary[];
  daily_summaries: StudyDailySummary[];
};

export async function fetchStudySubjects(): Promise<StudySubject[]> {
  const res = await apiClient.get("/study/subjects");
  return res.data;
}

export async function createStudySubject(input: StudySubjectInput): Promise<StudySubject> {
  const res = await apiClient.post("/study/subjects", input);
  return res.data;
}

export async function fetchStudyLogs(params?: { start_date?: string; end_date?: string; subject_id?: number | null }): Promise<StudyLog[]> {
  const res = await apiClient.get("/study/logs", { params });
  return res.data;
}

export async function createStudyLog(input: StudyLogInput): Promise<StudyLog> {
  const res = await apiClient.post("/study/logs", input);
  return res.data;
}

export async function updateStudyLog(id: number, input: Partial<StudyLogInput>): Promise<StudyLog> {
  const res = await apiClient.patch(`/study/logs/${id}`, input);
  return res.data;
}

export async function deleteStudyLog(id: number): Promise<void> {
  await apiClient.delete(`/study/logs/${id}`);
}

export async function fetchStudySummary(params?: { start_date?: string; end_date?: string }): Promise<StudySummary> {
  const res = await apiClient.get("/study/summary", { params });
  return res.data;
}
